site/source/docs/porting/Debugging.rst

```diff
@@ -4,128 +4,200 @@
+above, preserving DWARF inhibits some Binaryen optimizations. However DWARF has the advantage
+that it includes information about inlining, which can result in more accurate stack traces.
 
-.. note:: The more limited amount of debug information can also be enabled by specifying the :ref:`verbose output <debugging-emcc-v>` compiler flag (``emcc -v``).
+Emscripten includes a tool called ``emsymbolizer`` that can map wasm code addresses to sources
+using several different kinds of debug info, including DWARF (in wasm object or linked files)
+and source maps for line/column info, and symbol maps (see :ref:`emcc-emit-symbol-map`),
+name sections and object file symbol tables for function names.
 
+Examples:
+
+.. code-block:: bash
+
+  emcc source.c -c -o source.o -g # source.o has DWARF sections (-gsource-map also works here)
+  emcc source.o -o program.js -gsource-map # program.wasm.map contains a source map
+
+  emcc source.o -o program2.js -g # program2.wasm has DWARF
+  llvm-strip program2.wasm -o program2_stripped.wasm # program2_stripped.wasm has no debug info
 
-.. _debugging-compilation-settings:
 
-Compiler settings
-==================
+Fast Edit+Compile with minimal debug information
+================================================
 
-Emscripten has a number of compiler settings that can be useful for debugging. These are set using the :ref:`emcc -s<emcc-s-option-value>` option, and will override any optimization flags. For example:
+When you want the fastest builds, you generally want to avoid generating large debug information
+during compile, because it takes time to link into the final binary. It is still worthwhile to use
+the ``-g2``
+flag (at link time only) because browsers understand the name section even when devtools are not 
+in use, resulting in more useful stack traces at minimal cost.
+
+Example:
 
 .. code-block:: bash
 
-  emcc -O1 -sASSERTIONS test/hello_world
+  emcc source.c -c -o source.o # source.o has no debug info
+  emcc source.o -o program.js -g2 # program.wasm has a name section, program.js is unminified
+
+Sometimes the use of the ``-O1`` or ``-Og`` flag at compile time can also result in faster
+builds, because optimizations early in the pipeline can reduce the amount of IR that is
+processed by later phases such as instruction selection and linking. It also of course
+reduces test runtime.
+
+
+Detecting Memory Errors and Undefined Behavior
+==============================================
+
+The best tools for detecting memory safety and undefined behavior issues. are Clang's sanitizers,
+such as the Undefined Behaviour Sanitizer (UBSan) and the Address Sanitizer (ASan).
+For more information, see :ref:`Sanitizers`.
+
+
+Emscripten has several other compiler settings that can be useful for catching errors at runtime.
+These are set using the :ref:`emcc -s<emcc-s-option-value>` option, and will override any optimization flags (TODO is this true?). For example:
```

Comment:
@kripken do you know if this is true? I think maybe not, other than ASSERTIONS disabling metadce?